frontend/src/metabase/visualizations/components/settings/ChartSettingTableColumns/TableColumnPicker/utils.ts

```diff
@@ -0,0 +1,88 @@
+import { getColumnIcon } from "metabase/common/utils/columns";
+import * as Lib from "metabase-lib";
+import { findColumnSettingIndexForColumn } from "metabase-lib/queries/utils/dataset";
+import { getColumnKey } from "metabase-lib/queries/utils/get-column-key";
+import type {
+  DatasetColumn,
+  TableColumnOrderSetting,
+} from "metabase-types/api";
+
+import type { EditWidgetData } from "../types";
+
+import type { ColumnItem, DragColumnProps } from "./types";
+
+export function getColumnItems(
+  query: Lib.Query,
+  stageIndex: number,
+  columns: DatasetColumn[],
+  columnSettings: TableColumnOrderSetting[],
+): ColumnItem[] {
+  const columnItems = columns.map(legacyColumn => {
```

Comment:
Columns from query results are used for the source of truth, which is essentially the fix.